execute/aggregate.go

```diff
@@ -45,16 +45,18 @@ type AggregateTransformation interface {
 }
 
 type aggregateTransformation struct {
-	t AggregateTransformation
-	d *TransportDataset
+	t          AggregateTransformation
+	d          *TransportDataset
+	freeStates []interface{}
 }
 
 // NewAggregateTransformation constructs a Transformation and Dataset
 // using the aggregateTransformation implementation.
 func NewAggregateTransformation(id DatasetID, t AggregateTransformation, mem memory.Allocator) (Transformation, Dataset, error) {
 	tr := &aggregateTransformation{
-		t: t,
-		d: NewTransportDataset(id, mem),
+		t:          t,
+		d:          NewTransportDataset(id, mem),
+		freeStates: make([]interface{}, 10),
```

Comment:
This might also be causing the bug. When making a new slice with a capacity you need to specify both the initial length and desired capacity of the slice like so `make([]type, length, capacity)`. If you leave off the second value it behaves as if both length and capacity are the same. 

As a result we are creating a slice of states that starts with 10 entries all of which are nil. Since we only add new states if `len() < cap()` we only ever get to use one state slot (since we pop first before pushing).

Using `make([]interface{}, 0, 10)` will create an empty slice with capacity for 10 values.